tests/end_to_end/helpers.h

```diff
@@ -91,15 +93,95 @@ typedef class _program_information_provider
   public:
     _program_information_provider(ebpf_program_type_t program_type) : program_type(program_type)
     {
+        if (program_type == EBPF_PROGRAM_TYPE_XDP)
+            encode_xdp();
+        else if (program_type == EBPF_PROGRAM_TYPE_BIND)
+            encode_bind();
+        else
+            REQUIRE(program_type == EBPF_PROGRAM_TYPE_UNSPECIFIED);
+
         REQUIRE(
-            ebpf_provider_load(&provider, &program_type, nullptr, &provider_data, nullptr, nullptr, nullptr, nullptr) ==
-            EBPF_ERROR_SUCCESS);
+            ebpf_provider_load(
+                &provider,
+                &program_type,
+                nullptr,
+                reinterpret_cast<ebpf_extension_data_t*>(provider_data.data()),
+                nullptr,
+                nullptr,
+                nullptr,
+                nullptr) == EBPF_ERROR_SUCCESS);
     }
     ~_program_information_provider() { ebpf_provider_unload(provider); }
 
   private:
+    void
+    encode_bind()
+    {
+        ebpf_helper_function_prototype_t helper_functions[] = {
+            {1,
+             "ebpf_map_lookup_element",
+             EBPF_RETURN_TYPE_PTR_TO_MAP_VALUE_OR_NULL,
+             {EBPF_ARGUMENT_TYPE_PTR_TO_MAP, EBPF_ARGUMENT_TYPE_PTR_TO_MAP_KEY}},
+            {2,
+             "ebpf_map_update_element",
+             EBPF_RETURN_TYPE_INTEGER,
+             {EBPF_ARGUMENT_TYPE_PTR_TO_MAP, EBPF_ARGUMENT_TYPE_PTR_TO_MAP_KEY, EBPF_ARGUMENT_TYPE_PTR_TO_MAP_VALUE}},
+            {3,
+             "ebpf_map_delete_element",
+             EBPF_RETURN_TYPE_PTR_TO_MAP_VALUE_OR_NULL,
+             {EBPF_ARGUMENT_TYPE_PTR_TO_MAP, EBPF_ARGUMENT_TYPE_PTR_TO_MAP_KEY}},
+        };
+        ebpf_context_descriptor_t context_descriptor{
+            sizeof(bind_md_t), EBPF_OFFSET_OF(bind_md_t, app_id_start), EBPF_OFFSET_OF(bind_md_t, app_id_end), -1};
+        ebpf_program_type_descriptor_t program_type_descriptor{"bind", &context_descriptor};
+        ebpf_program_information_t program_information{
+            program_type_descriptor, _countof(helper_functions), helper_functions};
+        uint8_t* buffer;
+        unsigned long buffer_size;
+        REQUIRE(ebpf_program_information_encode(&program_information, &buffer, &buffer_size) == EBPF_ERROR_SUCCESS);
```

Comment:
does buffer need to be freed? (same question applies to line 176)

Reply:
@Alan-Jowett this question was marked resolved but I didn't see a change or a response here so unresolving

Reply:
Got it, thanks!